filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+package filewatcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+
+	"github.com/reddit/baseplate.go/internal/limitopen"
+)
+
+// Default option values
+const (
+	// DefaultFSEventsDelay is the default FSEventsDelay used when creating a new
+	// FileWatcher.
+	DefaultFSEventsDelay = 1 * time.Second
+
+	// DefaultPollingInterval is the default PollingInterval used when creating a
+	// new FileWatcher.
+	DefaultPollingInterval = 30 * time.Second
+
+	// DefaultInitialReadInterval is the default InitialReadInterval used when
+	// creating a new FileWatcher.
+	DefaultInitialReadInterval = time.Second / 2
+
+	// DefaultMaxFileSize is the default MaxFileSize used when it's <= 0.
+	//
+	// It's 10 MiB, with hard limit multiplier of 10.
+	DefaultMaxFileSize  = 10 << 20
+	HardLimitMultiplier = 10
+)
+
+// FileWatcher loads and parses data from a file or directory, and watches for
+// changes in order to refresh its stored data.
+type FileWatcher[T any] interface {
```

Comment:
It could be useful for users to have access to some timing data such as the last time the data was reloaded successfully.

Reply:
The parser hook seems like a reasonable option. Thanks.